extensions/ql-vscode/src/packaging.ts

```
@@ -0,0 +1,135 @@
+      step: 2,
+      maxStep: 2,
+    });
+    try {
+      await cliServer.packDownload(packsToDownload);
+      void showAndLogInformationMessage('Finished downloading packs.');
+    } catch (error) {
+      void showAndLogErrorMessage(
+        'Unable to download all packs. See logs for more details.'
+      );
+    }
+  }
+}
+
+interface QLPackQuickPickItem extends QuickPickItem {
+  packRootDir: string[];
+}
+
+/**
+ * Prompts user to choose packs to install, and installs them.
+ *
+ * @param cliServer The CLI server.
+ * @param progress A progress callback.
+ */
+export async function handleInstallPacks(
+  cliServer: CodeQLCliServer,
+  progress: ProgressCallback,
+): Promise<void> {
+  progress({
+    message: 'Choose packs to install',
+    step: 1,
+    maxStep: 2,
+  });
+  const workspacePacks = await cliServer.resolveQlpacks(getOnDiskWorkspaceFolders());
+  const quickPickItems = Object.entries(workspacePacks).map<QLPackQuickPickItem>(([key, value]) => ({
+    label: key,
+    packRootDir: value,
+  }));
+  const packsToInstall = await window.showQuickPick(quickPickItems, {
+    placeHolder: 'Select packs to install',
+    canPickMany: true,
+    ignoreFocusOut: true,
+  });
+  if (packsToInstall && packsToInstall.length > 0) {
+    progress({
+      message: 'Installing packs. This may take a few minutes.',
+      step: 2,
+      maxStep: 2,
+    });
+    const failedPacks = [];
+    const errors = [];
+    for (const pack of packsToInstall) {
+      try {
+        for (const dir of pack.packRootDir) {
+          await cliServer.packInstall(dir);
+        }
+      } catch (error) {
+        failedPacks.push(pack.label);
+        errors.push(error);
+      }
+    }
+    if (failedPacks.length > 0) {
+      void logger.log(`Errors:\n${errors.join('\n')}`);
+      void showAndLogWarningMessage(
+        `Unable to install some packs: ${failedPacks.join(', ')}. See logs for more details.`
+      );
```

Maybe here you can throw the error instead of just logging it. You can then structure your tests around that.